### ISSUE TITLE
Support custom short URL schemes

### ISSUE THREAD

**Issue by uqs** [vimwiki/vimwiki#1189]
Hi, so I'm on Stable 2.5 and am trying to use vimwiki for a sort of Zettelkasten. I would like to implement my own linkifying magic, so this is more a question about VimScript and why things don't quite work the way I think they should.

In `vars.vim` I found the regex that will automatically highlight http URLs, without the need to stick them in [[]] tags. Great, so I want to support things like b/123 to automatically link to our bug system's ticket number 123. So I extended the web_schemes with my own thing, but that alone doesn't do anything, I also had to extend rxWebProtocols in this file.

Question 1: As I didn't want to modify the stock sources, I figured I can override these regexen in `after/ftplugin/vimwiki.vim` like so:

```
" let g:vimwiki_global_vars.rxSchemes = '\%(wiki\d\+\|diary\|local\|http\|https\|file\|ftp\|gopher\|telnet\|nntp\|ldap\|rsync\|imap\|pop\|irc\|ircs\|cvs\|svn\|svn+ssh\|git\|ssh\|fish\|sftp\|mailto\|news\|xmpp\|sip\|sips\|doi\|urn\|tel\|data\|go\|b\|cl\|who\)'
let g:vimwiki_global_vars.rxWeblinkUrl = '\%(\%(\%(http\|https\|file\|ftp\|gopher\|telnet\|nntp\|ldap\|rsync\|imap\|pop\|irc\|ircs\|cvs\|svn\|svn+ssh\|git\|ssh\|fish\|sftp\):\%(//\)\)\|\%(mailto\|news\|xmpp\|sip\|sips\|doi\|urn\|tel\|data\):\|\%(go\|b\|cl\|who\)\/\)\S\{-1,}\%(([^ \t()]*)\)\='                               
let g:vimwiki_syntax_variables['default'].rxWeblink =
        \ '\<'. g:vimwiki_global_vars.rxWeblinkUrl . '[^[:space:]>]*'
let g:vimwiki_syntax_variables['default'].rxWeblinkMatchUrl =
        \ g:vimwiki_syntax_variables['default'].rxWeblink
let g:vimwiki_syntax_variables['default'].rxAnyLink =
        \ g:vimwiki_syntax_variables['default'].rxWikiLink.'\|'.
        \ g:vimwiki_global_vars.rxWikiIncl.'\|'.g:vimwiki_syntax_variables['default'].rxWeblink
```

And while the content then is "correct" after I file is loaded, these don't actually _do_ anything. Why can't I override this in that way?

So I have the following patch directly hacked in now:

```
diff --git autoload/vimwiki/vars.vim autoload/vimwiki/vars.vim
index f98ddef..e110932 100644
--- autoload/vimwiki/vars.vim
+++ autoload/vimwiki/vars.vim
@@ -43,11 +43,14 @@ function! s:populate_global_variables() abort
         \ 'git', 'ssh', 'fish', 'sftp'], '\|')
   let web_schemes2 =
         \ join(['mailto', 'news', 'xmpp', 'sip', 'sips', 'doi', 'urn', 'tel', 'data'], '\|')
+  let web_schemes3 =
+       \ join(['go', 'b', 'cl', 'who'], '\|')
 
   let g:vimwiki_global_vars.rxSchemes = '\%('.
         \ g:vimwiki_global_vars.schemes . '\|'.
         \ g:vimwiki_global_vars.web_schemes1 . '\|'.
-        \ web_schemes2 .
+        \ web_schemes2 . '\|' .
+        \ web_schemes3 .
         \ '\)'
 
   " match URL for common protocols; see http://en.wikipedia.org/wiki/URI_scheme
@@ -60,9 +63,14 @@ function! s:populate_global_variables() abort
           \ '\)'.
         \ '\|'.
           \ '\%('.web_schemes2.'\):'.
+        \ '\|'.
+          \ '\%('.web_schemes3.'\)\/'.
+        \ '\|'.
+          \ '\%([a-z]\+@\)'.
         \ '\)'
 
-  let g:vimwiki_global_vars.rxWeblinkUrl = rxWebProtocols . '\S\{-1,}'. '\%(([^ \t()]*)\)\='
+  "let g:vimwiki_global_vars.rxWeblinkUrl = rxWebProtocols . '\S\{-1,}'. '\%(([^ \t()]*)\)\='
+  let g:vimwiki_global_vars.rxWeblinkUrl = rxWebProtocols . '\%(([^ \t(),:]*)\)\='
 
   let wikilink_prefix = '[['
   let wikilink_suffix = ']]'
```

Question 2: I would also like to be able to turn `@user` into a link to our people/contact website, same for `user@`. But here using a protocol scheme of just `@` doesn't work, as the URL part then is empty ... Any ideas how to do this in a nice way?

Here's my handler in any case:

and of course:
```
function! VimwikiLinkHandler(link)
    if a:link =~# '\v^(go|who|b|cl)/'
        call vimwiki#base#system_open_link('http://' . a:link)
        return 1
    elseif a:link =~# '[a-z]*@'
        call vimwiki#base#system_open_link('http://who/' . matchstr(a:link, '\a\+'))
        return 1
    else
        return 0
    endif
endfunction
```

tl;dr can we have a setting where a user provides a set of _additional_ regexen that should be rendered as links and also trigger the linkhandler when hitting Enter?

**Comment by tinmarino** [comment]
Nice analysis. vars.vim is complicated to read / write.
You cannot hack it in a after directoy as it is used by syntax to set otehr internal variables. It is too late.

Please use the dev branch as vimwiki has a slow release cycle and many changes were made this year.

```
let g:vimwiki_global_vars['schemes_web'] = 'the regex you want'
```

Is what you are looking for.

Note that recently I added the command `:VimwikiVar` which can be helpfull to debug the user variables and their cobversion at loading vimwiki (present only in dev branch).
Any user variable must be set before `packadd vimwiki`.

Those exported variables may not be enought for your requirements. In that case, feel free to change the code exporting a generic string variable and make a PR (other resaon to use the dev branch).

I think the feature you are looking for is very interesting, it would help me too to jump to the tickets. I never took time to look as deep as you and just remapped `gx` to a custom url expander.
